[Brandon Ong] iP

botr99 opened the pull request:
# Duke
> A task management app built for me, for you and for everyone

Are you constantly **stressed** about the tasks you need to do?

Fret not, Duke is here to help you!

Why use Duke? Because it
- is easy to learn and use
- comes ~~with a subscription price of $10/month~~ **FREE** forever

The set-up is easy. Simply,
1. Download the jar file [here](https://github.com/botr99/ip/releases/tag/v0.1).
2. Copy the jar file into an *empty* folder.
3. Open a terminal at that folder.
4. Run `java -jar ip.jar` and you are good to go! 😎 

Features:
- [x] Managing todos
- [x] Managing events and deadlines
- [x] Saving of tasks to the hard disk at all times
- [x] Searching tasks by their descriptions
- [ ] GUI support (coming soon)   

Here is a short snippet of the `main` method:
```java
public class Duke {
    private static final String STORAGE_LOCATION = "data/duke.txt";

    public static void main(String[] args) {
        try {
            new Duke(STORAGE_LOCATION).run();
        } catch (IOException e) {
            System.out.println("Something went wrong with accessing " + STORAGE_LOCATION);
        } catch (FileParseException e) {
            System.out.println(STORAGE_LOCATION + " is not of the proper format. "
                    + "Either modify it to the right format or delete the file.");
        }
    }
}
```

Commit:
Add Gradle support

botr99 committed:
Change greet display format

botr99 committed:
Implement greet, echo, exit

botr99 committed:
Implement add, list

botr99 committed:
Implement Task class to represent tasks

botr99 committed:
Add the ability to mark tasks as done

botr99 committed:
Set up classes to distinguish multiple types of tasks

botr99 committed:
Implement todos, events, deadlines

botr99 committed:
Set up I/O redirection to semi-automate the testing of Duke

botr99 committed:
Add DukeException class to represent exceptions specific to Duke

botr99 committed:
Integrate DukeException and write more text UI tests

botr99 committed:
Replace fixed size tasks array with an ArrayList<Task>

botr99 committed:
Implement delete

botr99 committed:
Implement file creation, reading and writing

botr99 committed:
Add parsing of lines and updating of lines from a file

botr99 committed:
Update tasks to fit the file format

botr99 committed:
Implement saving tasks to data/duke.txt

botr99 committed:
Ignore the tasks saved to the user's data/ folder

botr99 committed:
Implement dates to deadlines and events

botr99 committed:
Merge branch 'branch-Level-7'

botr99 committed:
Merge branch 'branch-Level-8'

There are conflicts of behaviour now that Deadline and Event both throw
DukeException where it didn't beforehand.

This affects file parsing from storage since the Deadline and Event
created from reading the contents of the file now needs to handle the
DukeException.

Catching the DukeException and in turn throwing a custom
FileParseException allows us to keep file processing behaviour seperate
from actual app behaviour (i.e. Duke behaviour).

botr99 committed:
Extract Ui, Storage, Parser, TaskList and Message

Duke contains too many methods. Some of the methods are just for
interacting with the user; some are just loading and saving tasks to
file; some are just dealing with user inputs; some are involved with
manipulating the list of tasks; the rest is just printing out responses
made by the chatbot.

This causes the code to be hard to follow along.

Extracting the methods into their respective classes delegates
responsibilities to the classes and makes it easier to refactor certain
behaviour of Duke by just refactoring only the class that handles this
behaviour.

botr99 committed:
Fix tasks not getting saved correctly to storage

botr99 committed:
Fix issue of ui starting though tasks are not initialized

botr99 committed:
Put all classes in duke package

botr99 committed:
Edit file paths for README and text tests

botr99 committed:
Put tasks in duke.task package

botr99 committed:
Add equals method for Task and DateTask

botr99 committed:
Add JUnit tests

botr99 committed:
Set up the app as a JAR file

botr99 committed:
Add JavaDoc for tasks

botr99 committed:
Add JavaDoc for the remaining classes

botr99 committed:
Reduce line length to comply with coding standard

botr99 committed:
Add find tasks functionality

botr99 committed:
Merge branch 'branch-A-JavaDoc'

botr99 committed:
Merge branch 'branch-A-CodingStandard'

botr99 committed:
Merge branch 'branch-Level-9'

botr99 committed:
Refactor TaskList and Storage

botr99 committed:
Update tests

botr99 committed:
Merge remote-tracking branch 'origin/add-gradle-support' into branch-A-Gradle

botr99 committed:
Set up Gradle

botr99 committed:
Modify test file location to work with Gradle

botr99 committed:
Merge branch 'branch-A-Gradle'

botr99 committed:
Configure Checkstyle rules

GraceWang2322 commented on src/main/java/duke/DukeException.java:
```diff
@@ -0,0 +1,17 @@
+ *
+ * @author botr99
+ */
+public class DukeException extends Exception {
```
maybe add a line before starting on the header comment for Duke Exception constructor?

GraceWang2322 commented on src/main/java/duke/FileParseException.java:
```diff
@@ -0,0 +1,19 @@
+ * @author botr99
+ */
+public class FileParseException extends Exception {
+    /**
```
maybe can add a blank line in here as well

GraceWang2322 commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,91 @@
+
+import java.io.IOException;
+
+import java.nio.file.Files;
```
i like how you organised the import command lines according to their usage 👍

GraceWang2322 reviewed (commented):
It's really nice hw ur code is super clean. The javadoc and coding standard is well-followed as well. 👍 💯

botr99 committed:
Modify files to pass CheckStyle

botr99 committed:
Merge branch 'branch-A-CheckStyle'

stanley-1 commented on src/main/java/duke/FileParseException.java:
```diff
@@ -0,0 +1,19 @@
+package duke;
+
+/**
+ * Represents an exception that occurs when the
```
Perhaps you can consider editing your JavaDoc comment to the following for better phrasing: "Throws an Exception when the file format does not conform with the expected String representation of a Task." 😄

stanley-1 commented on src/main/java/duke/Message.java:
```diff
@@ -0,0 +1,119 @@
+     *
+     * @param tasks The user's task list.
+     */
+    public static void printTasksMessage(TaskList tasks) {
```
For added readability, you can consider renaming the method to "printTasklist(...)".

stanley-1 commented on src/main/java/duke/Message.java:
```diff
@@ -0,0 +1,119 @@
+     *
+     * @param task The task that was marked as done.
+     */
+    public static void printMarkTaskDoneMessage(Task task) {
```
For enhanced readability, you can consider renaming this method to "printMarkTaskAsDoneMessage(...)".

stanley-1 reviewed (commented):
Hi Brandon! I am impressed with the meticulousness of your coding quality and choice of names for your various methods and classes. In addition, you've provided a variety of tests to evaluate the functionality of your application. You've done an impeccable job overall and listed below are just some suggestions which you can consider implementing. 😄

botr99 committed:
Set up GUI for Duke

botr99 committed:
Merge branch 'branch-Level-10'

reddevil1313 commented on README.md:
```diff
@@ -13,12 +13,15 @@ Prerequisites: JDK 11, update Intellij to the most recent version.
    1. If there are any further prompts, accept the defaults.
```
Would suggest adding more information in your README on how your software works and more clear steps on how one should download it

reddevil1313 commented on src/main/java/duke/Gui.java:
```diff
@@ -0,0 +1,24 @@
+    }
+
+    /**
+     * {@inheritDoc}
```
Nice use of @inheritDoc

reddevil1313 commented on src/main/java/duke/task/Event.java:
```diff
@@ -0,0 +1,62 @@
+package duke.task;
```
Maybe your javadocs are a bit too long, can summarise better?

reddevil1313 reviewed (commented):
Overall good job! Just small improvements in documentation required!

botr99 committed:
Group related classes into packages

There are many related classes belonging to the root duke package.

These related classes cause duke package to be congested and hard to
follow along.

Let's group DukeException and FileParseException into an exception
package within the duke package. Similarly, let's group Storage and
TaskStorage into a storage package. Finally, let's group Cli, Gui
and Ui into a ui package.

This makes it easier to find exception-related code, storage-related
code and ui-related code respectively.

botr99 committed:
Add assertions

Assertions help to define assumptions about the program state so that
the runtime can verify them. These definitions are useful to programmers
working on the repo in future.

Let's verify that
- the file we are reading the user's data from exists after a storage is
constructed to store the file content.
- the file content stored in the storage is up-to-date with changes made
to the file.
- for tasks that consists of a date, these tasks have a date after being
constructed.

botr99 committed:
Rename all variables of type TaskList

For variables of type TaskList, tasks is mainly used.

Let's rename all these variables named tasks to taskList.

tasks is mainly used to refer to ArrayList<Task>, and
TaskList behaves similarly to ArrayList<Task>. Renaming tasks
to taskList for variables of type TaskList helps reduce
confusion between the two.

botr99 committed:
Refactor Message

botr99 committed:
Seperate commands into individual classes

botr99 committed:
Refactor Parser and update ParserTest

Ui is handling user input by parsing it.

It is not the job of ui to do so.

Let's delegate this function to Parser instead, and take advantage of
the newly created Command classes.

botr99 committed:
Shift getResponse from Duke to Ui

This makes it so that it is always invoked regardless of Duke being
run as a command line application or as a gui.

botr99 committed:
Merge pull request #1 from botr99/branch-A-Assertions

Add assertions

botr99 committed:
Merge branch 'master' of https://github.com/botr99/ip into branch-A-CodeQuality

botr99 committed:
Merge pull request #2 from botr99/branch-A-CodeQuality

Improve code quality

botr99 committed:
Set up CI

botr99 committed:
Set up reminders

botr99 committed:
Merge branch 'branch-B-Reminders'

botr99 committed:
Improve GUI

- Set spacing between image and text
- Make the image rounded
- Add a background colour for messages made by Duke

botr99 committed:
Add product screenshot

botr99 committed:
Add user guide

botr99 committed:
Set theme jekyll-theme-slate

botr99 committed:
Prepare for v0.2 release

botr99 committed:
Fix typo in jar file link

botr99 committed:
Edit user guide
